src/plugins/console/README.md

```diff
@@ -2,4 +2,55 @@
 
 ## About
 
-Console provides the user with tools for storing and executing requests against Elasticsearch.
\ No newline at end of file
+Console provides the user with tools for storing and executing requests against Elasticsearch.
+
+## Architecture
+- Ace editor wrapped with `<KIBANA_REPO>/src/plugins/console/public/types/core_editor.ts`. This wrapper allows for an easier change of the editor, for example from Ace to Monaco.
+- Autocomplete logic in `<KIBANA_REPO>/src/plugins/console/public/lib/autocomplete`. Autocomplete rules are computed by classes in `components` folder.
+
+## Autocomplete definitions
+Users benefit greatly from autocomplete suggestions since not all Elasticsearch APIs can be provided with a corresponding UI. The autocomplete definitions are all created in the form of javascript objects loaded from `json` or `js` files. 
+
+### Creating definitions
+The folder `<KIBANA_REPO>/src/plugins/console/server/lib/spec_definitions/json/generated` contains definitions generated automatically from Elasticsearch REST API specifications. See `<KIBANA_REPO>/packages/kbn-spec-to-console/README.md` for more information on the `spec-to-console` script. 
+
+Manually created override files in the folder `<KIBANA_REPO>/src/plugins/console/server/lib/spec_definitions/json/overrides` contain fixes for generated files and additions for body request parameters.   
+
+### Top level keys
+Following top level keys are used in the definitions objects:
+
+- `documentation`: string for the docs url link, `master` and `current` strings in the url are replaced with the `docLinkVersion`
+- `methods`: array of strings for allowed http methods
+- `patterns`: array of strings for url endpoints, accept parameters like `{indices}`, `{fields}` etc
+- `url_params`: an object with possible keys for url parameters and its values
+- `priority`: a number to decide when there are several possible autocomplete definitions
+- `data_autocomplete_rules`: an object with possible keys for body request and its values
+
+### Url parameters
+When defining url parameters or body request parameters, each key in the object is a possible url parameter or a possible field in the body request. Parameters like `{indices}`, `{fields}` etc are accepted both as a key and a value.
+
+- `__flag__`: is used for boolean params, suggestions contain `true` and `false`
+- array with a list of values: is used for a list of known accepted param values, suggestions contain all values in the list
+
+### Body request params:
+- `__one_of: [..., ...]`: is used for a list of known accepted param values, autocompletes as the first value of the array, suggestions contain all values in the list
```

Comment:
Bullets work well for short lists. But I've noticed that they stop working once they meet a couple criteria:

* When the number of bullets grows to 6 or 7
* When the content of each bullet becomes so large that I am forced to sacrifice complete sentences for sentence fragments

Once I notice these symptoms, I've found it better to switch from bullets to dedicated sections. So for this list, I'd create a section for each body request param type, which will give me the space I need to write full sentences, provide detailed examples, and express anything else I need that will communicate my ideas. Example below.

---

#### __one_of

`__one_of` defines a list of accepted values for a body field. It will autocomplete to the first value in the array. The autocomplete suggestion list will render all of the items in the defined array.

```js
my_field: {
  __one_of: ['value1', 'value2', 3, 4, 5],
},
```

For a boolean field, use `[true, false]`.

#### __any_of

`__any_of` behaves the save as `__one_of`, except that it...